### PR TITLE
Implement the Go PCLN Table Parser

### PR DESCRIPTION
**Pull request by tommyknows**
- [x] Ready for review
- [x] Follows CONTRIBUTING rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?

This commit / PR  adds a PCLN table parser to read from a Go binary and get
all files that were used to build that binary.
It is a requirement to read the packages from stripped Go binaries.

See [MYC-145](https://snyksec.atlassian.net/browse/MYC-145).

The main parsing code has been copied from the Go source and adapted to
our needs.

It does not integrate with anything yet, that work is tracked in
[MYC-146](https://snyksec.atlassian.net/browse/MYC-146).

#### How should this be manually tested?

One thing that would be good is if the binary-generating script could be run to make sure it's actually reproducible.

#### What are the relevant tickets?

[MYC-145](https://snyksec.atlassian.net/browse/MYC-145).

### CODE REVIEW

File: lib/go-parser/symbols/index.ts
Comment (yaronschwimmer) [inline lib/go-parser/symbols/index.ts:178]: will future Go versions require us to modify this code?

File: lib/go-parser/symbols/index.ts
Comment (yaronschwimmer) [inline lib/go-parser/symbols/index.ts:106]: I know this is basically a JS translation to a golang code, but this type of variable naming, though common in Go, is not considered best practice in JS. Please try to name variables in a more meaningful way

File: test/unit/go-version-parser.spec.ts
Comment (noaleibo1) [inline test/unit/go-version-parser.spec.ts:15]: Not sure I understand why we need this if. If one of us accidentally added something strange to `../fixtures/go-binaries` I'm fine with the tests breaking in runtime 🤷‍♀️

File: test/fixtures/go-binaries/symbols/source/build.sh
Comment (noaleibo1) [inline test/fixtures/go-binaries/symbols/source/build.sh:1]: A thought - do you think that we need to have this script as part of our `build` script, and remove the binary files from git?

**Review by noaleibo1 — COMMENTED** [review]
Nice, clean code - love it!
I had some comments, let me know what you think.
Also - I agree with Yaron's comment about the naming of variables.

File: lib/go-parser/symbols/index.ts
Comment (tommyknows) [reply lib/go-parser/symbols/index.ts:178]: Potentially, yes. I wanted to look into the cronjob you added that builds binaries with the latest version & checks compatibility, and integrate this parser into these tests as well. In a separate PR, probably :)

File: test/unit/go-version-parser.spec.ts
Comment (tommyknows) [reply test/unit/go-version-parser.spec.ts:15]: The tests already break because I've added a `symbols` folder, that's why we need it :-)

File: test/fixtures/go-binaries/symbols/source/build.sh
Comment (tommyknows) [reply test/fixtures/go-binaries/symbols/source/build.sh:1]: I don't think that makes a lot of sense as this takes a "long" time to build (~5m). Also, the binaries should be reproducible, so there's no real motivation to rebuild them again. I've kept this more as a reference for future builds (with newer Go versions) and that we know what is in these binaries :-)

File: test/unit/symbol-parser.spec.ts
Comment (tommyknows) [reply test/unit/symbol-parser.spec.ts:9]: Not sure what to add there without duplicating information or having it go out of date. What would you expect on the notion page?

File: lib/go-parser/symbols/index.ts
Comment (tommyknows) [reply lib/go-parser/symbols/index.ts]: Done 👍(I think 😄)

**Comment by tommyknows** [comment]
Manually triggered the Go binary regression test [here](https://app.circleci.com/pipelines/github/snyk/snyk-docker-plugin/1788/workflows/b566d4dc-a57c-4ca2-867d-3a7c6aac90e4/jobs/8671).

Decided to not merge the parsing code with the other information-grabbing code yet, seems to be too much work while [MYC-146](https://snyksec.atlassian.net/browse/MYC-146) is not implemented yet. Will do it there.

@yaronschwimmer / @noaleibo1 can you have another look & let me know if this looks good to you? Thanks!

File: test/fixtures/go-binaries/source/build.sh
Comment (yaronschwimmer) [inline test/fixtures/go-binaries/source/build.sh:8]: `latest` is not here

File: test/fixtures/go-binaries/source/build.sh
Comment (yaronschwimmer) [inline test/fixtures/go-binaries/source/build.sh:16]: shouldn't this condition be `==`?

File: test/fixtures/go-binaries/source/main.go
Comment (yaronschwimmer) [inline test/fixtures/go-binaries/source/main.go:1]: this file looks familiar :)
was it copied as is? can use it only once?

File: test/fixtures/go-binaries/source/build.sh
Comment (tommyknows) [reply test/fixtures/go-binaries/source/build.sh:16]: TIL: `==` is bash, `=` is POSIX shell. 

This is a shell-script, not a bash-script, so no :-)

File: test/fixtures/go-binaries/source/build.sh
Comment (tommyknows) [reply test/fixtures/go-binaries/source/build.sh:8]: Deliberately, I don't think we want to add `latest` binaries to the checked in files. Not having latest in here means we always build concrete versions, and if the regression-test should fail we can add the failing versions in here and have that built explicitly again.

File: test/fixtures/go-binaries/source/main.go
Comment (tommyknows) [reply test/fixtures/go-binaries/source/main.go:1]: Copied as is. We don't need the files in `test/go-binary-regression` anymore, so I guess I need to remove those still :)

That brings up another question: should I move this `source` folder to `test/go-binary-regression` instead? Imo keeping it in the `fixtures/go-binaries` directory is clearer as this is the place I'm searching for the source files as well. But YMMV, and I don't have a strong opinion on this :)

File: test/fixtures/go-binaries/source/build.sh
Comment (yaronschwimmer) [reply test/fixtures/go-binaries/source/build.sh:16]: I did not know that :)

File: test/fixtures/go-binaries/source/main.go
Comment (yaronschwimmer) [reply test/fixtures/go-binaries/source/main.go:1]: It makes more sense where it is with changes you made. Just remove the duplicated files from `go-binary-regression`

File: test/unit/go-binary-information-parser.spec.ts
Comment (yaronschwimmer) [inline test/unit/go-binary-information-parser.spec.ts:194]: you might want to read about `expect.arrayContaining` 
https://jestjs.io/docs/expect#expectarraycontainingarray

Comment (tommyknows) [reply test/unit/go-binary-information-parser.spec.ts:194]: Nice one, I wondered if there wasn't a better way to do this :smile: Thanks!

File: test/fixtures/go-binaries/source/main.go
Comment (tommyknows) [reply test/fixtures/go-binaries/source/main.go]: Done.

File: lib/go-parser/pclntab.ts
Comment (yaronschwimmer) [inline lib/go-parser/pclntab.ts:136]: can you add return type for this function?

Comment (tommyknows) [reply lib/go-parser/pclntab.ts:136]: I certainly can... I thought not writing anything means we don't return anything... 🙃

**Review by yaronschwimmer — APPROVED** [review]

File: test/unit/go-version-parser.spec.ts
Comment (noaleibo1) [reply test/unit/go-version-parser.spec.ts]: Ah I see, alright then.

File: test/fixtures/go-binaries/symbols/source/build.sh
Comment (noaleibo1) [reply test/fixtures/go-binaries/symbols/source/build.sh]: Alright, I'm fine with that.

File: test/unit/symbol-parser.spec.ts
Comment (noaleibo1) [reply test/unit/symbol-parser.spec.ts]: I'm not worried about duplicating information, I just want a link we can send on call/support/customer success when they have a question about the support to cgo and stripped binaries. What do you think about adding it to the existing spike document?

Comment (tommyknows) [reply test/unit/symbol-parser.spec.ts]: I think I'll add something to the docs once this is merged, so that I can link both PRs as well. Or do you think I need to document something already?

File: test/fixtures/go-binaries/symbols/source/build.sh
Comment (tommyknows) [reply test/fixtures/go-binaries/symbols/source/build.sh]: Yeah me neither, but I think it's the easiest & fastest thing to do... :/

**Pull request merged**
